wallet/src/Dscp/Wallet/KeyStorage.hs

```diff
@@ -0,0 +1,63 @@
+module Dscp.Wallet.KeyStorage
+       ( getAccounts
+       , addAccount
+       ) where
+
+import Data.Aeson (eitherDecode, encode)
+import Data.Aeson.TH (defaultOptions, deriveJSON)
+import Dscp.Core (mkAddr)
+import System.Directory (doesFileExist)
+import System.FileLock (SharedExclusive (..), withFileLock)
+
+import qualified Data.ByteString.Lazy as LBS
+
+import Dscp.Wallet.Face
+
+data Storage = Storage
+    { accounts :: [StorageAccount]
+    }
+
+data StorageAccount = StorageAccount
+    { secretKey :: Encrypted SecretKey
+    , publicKey :: PublicKey
+    }
+
+deriveJSON defaultOptions ''Storage
+deriveJSON defaultOptions ''StorageAccount
+
+storageFileName :: FilePath
+storageFileName = "wallet.json"
+
+storageLockFileName :: FilePath
+storageLockFileName = "wallet.json.lock"
+
+getAccounts :: IO [Account]
+getAccounts = withFileLock storageLockFileName Shared $ \_ -> do
+    storage <- readStorage
+    return $ toAccount <$> accounts storage
+  where
+    toAccount StorageAccount{..} = Account
+        { accountSecretKey = secretKey
+        , accountPublicKey = publicKey
+        , accountAddress = mkAddr publicKey
+        }
+
+addAccount :: Account -> IO ()
+addAccount account = withFileLock storageLockFileName Exclusive $ \_ -> do
+    Storage accs <- readStorage
+    writeStorage $ Storage $ accs ++ [fromAccount account]
+  where
+    fromAccount Account{..} = StorageAccount
+        { secretKey = accountSecretKey
+        , publicKey = accountPublicKey
+        }
+
+readStorage :: IO Storage
+readStorage = do
+    exists <- doesFileExist storageFileName
+    if exists
+        then LBS.readFile storageFileName >>= either fail return . eitherDecode
+        else return $ Storage []
+
+writeStorage :: Storage -> IO ()
+writeStorage = LBS.writeFile storageFileName . encode
```

Comment:
With this approach `withFileLock` should be used in each function which knows `Storage` internals. I'd move it to primitive functions (`readStorage`, `writeStorage` can be changed to `modifyStorage`), because number of these functions is definitely constant, while number of less primitive functions can increase over time (you may want to delete an account, reorder accounts, maybe add something else to the storage).

Reply:
Fair.

Reply:
Yes.